src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -998,5 +999,107 @@ static void RefParam(ref dynamic p)
                 expectedOutput: "frog",
                 additionalRefs: new[] { SystemCoreRef, CSharpRef }).VerifyDiagnostics();
         }
+
+        [Fact]
+        public void RefQueryClause()
+        {
+            // a "ref" may not precede the expression of a query clause...
+            // simply because the grammar doesn't permit it. Here we check
+            // that the situation is diagnosed, either syntactically or semantically.
+            // The precise diagnostics are not important for the purposes of this test.
+            var text = @"
+class C
+{
+    static void Main(string[] args)
+    {
+        var a = new[] { 1, 2, 3, 4 };
+        bool b = true;
+        int i = 0;
+        { var za = from x in a select ref x; } // error 1
+        { var zc = from x in a from y in ref a select x; } // error2
+        { var zd = from x in a from int y in ref a select x; } // error 3
+        { var ze = from x in a from y in ref a where true select x; } // error 4
+        { var zf = from x in a from int y in ref a where true select x; } // error 5
+        { var zg = from x in a let y = ref a select x; } // error 6
+        { var zh = from x in a where ref b select x; } // error 7
+        { var zi = from x in a join y in ref a on x equals y select x; } // error 8 (not lambda case)
+        { var zj = from x in a join y in a on ref i equals y select x; } // error 9
+        { var zk = from x in a join y in a on x equals ref i select x; } // error 10
+        { var zl = from x in a orderby ref i select x; } // error 11
+        { var zm = from x in a orderby x, ref i select x; } // error 12
+        { var zn = from x in a group ref i by x; } // error 13
+        { var zo = from x in a group x by ref i; } // error 14
+    }
+    public static T M<T>(T x, out T z) => z = x;
```

Comment:
Without looking closely at the issue, do you believe you are commenting on something that undermines the value of the test, or do you believe the test correctly tests for what it is intended to test?